File: ios/WatTracker/WatTracker/Screens/PairingFailureMessage.swift
import Foundation

/// What the rider is told when a pairing attempt does not work, and the one
/// thing this file exists to guarantee: that a wrong code, an expired code and
/// an already-used code produce the *same sentence*.
///
/// The server goes to some trouble to make those three indistinguishable --
/// `DevicePairingRegistry` in `wattracker/cloud/security.py` answers a
/// redemption failure the same way whatever the reason, and the read plane
/// answers every authentication failure with 404 for the same purpose. A UI
/// that renders `CloudClient.Failure` verbatim hands that back: "HTTP 410"
/// against one code and "HTTP 404" against another is an oracle, and an
/// attacker holding a screen full of guesses would rather have it than not.
/// So the default here is the shared message, and a condition earns a
/// distinct one only by being *provably independent of the code's value*.
///
/// Which conditions clear that bar:
///
/// - **No connection.** The request never reached a server, so the answer
///   cannot depend on what was in it.
/// - **The server asked us to wait** (429, 503). Admission control runs before
///   the pairing registry is consulted; the same refusal arrives for a valid
///   code, and telling the rider to wait rather than to find a new code is the
///   difference between the app being usable and not.
/// - **This device could not sign / could not read the reply.** Both are
///   statements about this client, not about the code, and both need a
///   different action from the rider than "get a new code".
/// - **The device's clock is off.** Also a fact about the device. It cannot
///   arise from `pair` today -- pairing carries no signature, so no timestamp
///   is checked -- but `CloudSession.Failure` can carry it and falling through
///   to a message about the code would be a lie if it ever did.
///
/// Everything else, including *every* HTTP status that is not 429 or 503,
/// lands on `codeRefused`. That is the fail-safe direction: a new condition
/// nobody anticipated becomes one more thing that looks like a bad code,
/// rather than one more thing that leaks.
enum PairingFailureMessage {
    /// The one message. Wrong, expired, already used, redeemed on another
    /// phone, never minted at all -- all of it reads exactly like this.
    static let codeRefused =
        "That code did not work. Ask the desktop app for a new one and type it in again."

    static func text(for error: Error) -> String {
        if let failure = error as? CloudSession.Failure {
            return text(for: failure)
        }
        // `CloudSession.pair` wraps everything, so a bare client failure only
        // reaches here from a caller that skipped the session. Mapped the same
        // way regardless, so the two paths cannot say different things about
        // the same refusal.
        if let failure = error as? CloudClient.Failure {
            return text(for: failure)
        }
        if error is URLError { return offline }
        return codeRefused
    }

    private static let offline =
        "No connection. Check that this device is on the same network as the desktop, "
        + "then try again."

    private static func text(for failure: CloudSession.Failure) -> String {
        switch failure {
        case .offline:
            return offline
        case let .throttled(retryAfter):
            return busy(retryAfter: retryAfter)
        case let .clockSkew(seconds):
            return "This device's clock is \(Int(abs(seconds)))s away from the server's. "
                + "Set the date and time automatically, then try again."
        case let .server(underlying):
            return text(for: underlying)
        case .notPaired, .deviceRemoved:
            // Reachable only if the credential changed underneath this attempt
            // -- another pairing landed first, or the session was wiped
            // mid-request. Nothing about the code follows from either.
            return codeRefused
        }
    }

    private static func text(for failure: CloudClient.Failure) -> String {
        switch failure {
        case let .http(status, _, retryAfter, _):
            // 429 and 503 are the only statuses the pairing route can return
            // without having looked at the code. Every other status -- 400,
            // 403, 404, 409, 410, 422, a 500 from a server bug -- collapses
            // into the shared message on purpose.
            guard status == 429 || status == 503 else { return codeRefused }
            return busy(retryAfter: retryAfter)
        case .nonceUnavailable:
            return "This device could not generate a signing nonce. Restart the app and try again."
        case .malformedResponse:
            // The pairing route's error bodies never reach here -- a non-2xx
            // becomes `.http` first -- so this is a 2xx this build could not
            // decode, which is a version mismatch and not a code.
            return "The server's reply could not be read. This app may need updating."
        case .noProfilePublished:
            // Not reachable from pairing; folded in rather than given a
            // sentence of its own, because the fail-safe default is the point.
            return codeRefused
        }
    }

    private static func busy(retryAfter: TimeInterval?) -> String {
        guard let retryAfter, retryAfter > 0 else {
            return "The server is busy. Wait a moment and try again."
        }
        return "The server is busy. Try again in \(Int(retryAfter.rounded(.up)))s."
    }
}

File: android/app/src/main/java/com/wattracker/android/cloud/PairingFailureMessage.kt
package com.wattracker.android.cloud

import java.io.IOException
import javax.net.ssl.SSLException
import kotlin.math.abs
import kotlin.math.ceil

/**
 * Maps errors during cloud pairing to user-facing failure messages.
 *
 * Guaranteed security property: wrong, expired, and already-used codes produce
 * the exact same sentence so the client does not act as an oracle. Conditions
 * earn distinct messages only if they are provably independent of the code's value
 * (e.g., offline connection, server throttled 429/503, clock skew).
 */
object PairingFailureMessage {

    const val CODE_REFUSED =
        "That code did not work. Ask the desktop app for a new one and type it in again."

    const val OFFLINE =
        "No connection. Check that this device is on the network, then try again."

    fun text(error: Throwable): String {
        return when (error) {
            is CloudSession.Failure -> textForCloudSessionFailure(error)
            is CredentialStoreException ->
                "Could not save pairing on this device. Restart the app and try again."
            is IOException -> OFFLINE
            else -> CODE_REFUSED
        }
    }

    private fun textForCloudSessionFailure(failure: CloudSession.Failure): String = when (failure) {
        is CloudSession.Failure.Offline -> OFFLINE
        is CloudSession.Failure.Throttled -> busy(failure.retryAfter)
        is CloudSession.Failure.ClockSkew ->
            "This device's clock is ${abs(failure.seconds).toInt()}s away from the server's. " +
                "Set the date and time automatically, then try again."
        is CloudSession.Failure.Server -> busyOrRefused(failure.status, null)
        is CloudSession.Failure.NotPaired, is CloudSession.Failure.DeviceRemoved -> CODE_REFUSED
    }

    private fun busyOrRefused(status: Int, retryAfter: Double?): String {
        return if (status == 429 || status == 503) {
            busy(retryAfter)
        } else {
            CODE_REFUSED
        }
    }

    private fun busy(retryAfter: Double?): String {
        return if (retryAfter != null && retryAfter > 0) {
            "The server is busy. Try again in ${ceil(retryAfter).toInt()}s."
        } else {
            "The server is busy. Wait a moment and try again."
        }
    }
}

/**
 * Maps errors during local desktop pairing to clear, actionable messages.
 */
object LocalPairingFailureMessage {

    fun text(error: Throwable): String {
        return when (error) {
            is LocalClientException.InsecureOrInvalidBaseUrl ->
                error.message ?: "Enter an HTTPS desktop address (e.g. https://192.168.1.10:8000)."
            is LocalClientException.MissingToken ->
                "Enter the connector token from the desktop web settings."
            is LocalClientException.Unauthorized ->
                "The desktop token was refused or has been revoked. Check the token in web settings and try again."
            is LocalClientException.UnexpectedLanding, is LocalClientException.MalformedResponse ->
                "The desktop's reply could not be read. This app may need updating."
            is LocalClientException.Http -> {
                if (error.status == 429 || error.status == 503) {
                    "The desktop server is busy (HTTP ${error.status}). Try again in a moment."
                } else {
                    "HTTP ${error.status} from desktop server."
                }
            }
            is SSLException ->
                "The desktop's HTTPS certificate is not trusted by this device."
            is IOException ->
                "No connection. Check that this device is on the same network as the desktop server."
            is CredentialStoreException ->
                "Could not save local credentials on this device. Restart the app and try again."
            else ->
                error.message ?: "Could not connect to the local desktop server."
        }
    }
}

/**
 * Result of attempting to remove a cloud device.
 */
sealed class RemoveDeviceResult {
    object ServerRevoked : RemoveDeviceResult()
    object LocalFallback : RemoveDeviceResult()
    data class Failed(val message: String) : RemoveDeviceResult()
}

object RemoveDeviceFailureMessage {

    fun text(error: Throwable): String {
        return when (error) {
            is CloudSession.Failure.Offline, is IOException ->
                "No connection. Check your network connection and try again."
            is CloudSession.Failure.Throttled ->
                "The server is busy. Try again in a moment."
            is CloudSession.Failure.Server ->
                if (error.status == 429 || error.status == 503) {
                    "The server is busy (HTTP ${error.status}). Try again in a moment."
                } else {
                    "Could not remove device (HTTP ${error.status})."
                }
            else ->
                error.message ?: "Could not remove device."
        }
    }
}
